test/py/CMakeLists.txt

```diff
@@ -23,37 +23,62 @@
 #####################################################################################
 
 include(PythonModules)
+set(VENV ${CMAKE_BINARY_DIR}/test/py/venv)
+set(BASE_REQUIREMENTS ${CMAKE_CURRENT_SOURCE_DIR}/base-requirements.txt)
+set(REQUIREMENTS_PYTHON_ALL ${CMAKE_CURRENT_SOURCE_DIR}/requirements-python-all.txt)
+set(REQUIREMENTS_PYTHON_36 ${CMAKE_CURRENT_SOURCE_DIR}/requirements-python-3.6.txt)
 
-function(add_py_test NAME SCRIPT)
+function(add_py_venv_fixture FIXTURE_NAME PYTHON_EXECUTABLE VIRTUAL_ENV_DIR REQUIREMENTS_FILE)
+    if(NOT TEST py_${FIXTURE_NAME}_initialize_env)
+        add_test(NAME py_${FIXTURE_NAME}_initialize_env COMMAND ${PYTHON_EXECUTABLE} -m venv ${VIRTUAL_ENV_DIR} --clear)
+        set_tests_properties(py_${FIXTURE_NAME}_initialize_env PROPERTIES FIXTURES_SETUP ${FIXTURE_NAME}_INIT_VENV)
+        set(ENV{VIRTUAL_ENV} ${VIRTUAL_ENV_DIR})
+        set(PYTHON_EXECUTABLE ${VIRTUAL_ENV_DIR}/bin/python)
+        add_test(
+            NAME py_${FIXTURE_NAME}_setup_env
+            COMMAND ${PYTHON_EXECUTABLE} -m pip install -r ${BASE_REQUIREMENTS} -r ${REQUIREMENTS_FILE})
+        set_tests_properties(py_${FIXTURE_NAME}_setup_env PROPERTIES FIXTURES_REQUIRED ${FIXTURE_NAME}_INIT_VENV)
+        set_tests_properties(py_${FIXTURE_NAME}_setup_env PROPERTIES FIXTURES_SETUP ${FIXTURE_NAME}_VENV)
+    endif()
+endfunction()
+
+function(add_py_test NAME SCRIPT VENV_DIR)
     foreach(PYTHON_VERSION ${PYTHON_VERSIONS})
         set (ENV_COMMAND ${CMAKE_COMMAND} -E env 
             "PYTHONPATH=$<TARGET_FILE_DIR:migraphx_pybind_${PYTHON_VERSION}>"
             "PYTHONMALLOC=debug"
             "MALLOC_CHECK_=3"
         )
         set(PYTHON_EXECUTABLE ${PYTHON_${PYTHON_VERSION}_EXECUTABLE})
+        if(${PYTHON_VERSION} STREQUAL "3.6")
+            add_py_venv_fixture(${PYTHON_VERSION} ${PYTHON_EXECUTABLE} ${VENV_DIR}/${PYTHON_VERSION} ${REQUIREMENTS_PYTHON_36})
+        else()
+            add_py_venv_fixture(${PYTHON_VERSION} ${PYTHON_EXECUTABLE} ${VENV_DIR}/${PYTHON_VERSION} ${REQUIREMENTS_PYTHON_ALL})
```

Comment:
The `add_py_env_fixture` probably needs to be called outside of the function since tests will use different virtual environments.

Reply:
I have taken the function out of `add_py_test`.